A32-xds-circuit-breaking.md

```diff
@@ -0,0 +1,157 @@
+implemented for [traffic splitting and routing](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md),
+each CDS LB policy uniquely represents the load balancing logic for all requests
+sent to the upstream cluster.
+
+## Overview
+
+gRPC client receives circuit breaking configurations as part of the [Cluster](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster.proto#cluster) 
+resource configuration in Cluster Discovery Service (CDS) responses. Each
+circuit breaker [Thresholds](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster/circuit_breaker.proto#cluster-circuitbreakers-thresholds)
+defines settings for a RoutingPriority. gRPC only takes the first 
+_Thresholds_ with `priority` being [`DEFABUT`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/base.proto#enum-core-routingpriority).
+
+As of the current gRPC design, only the `max_requests` field, which limits the 
+maximum number of requests that can be outstanding to all hosts in a cluster 
+at any given time, in `Thresholds` is considered. All the others are either 
+inapplicable to gRPC or the corresponding features are missing for now. The
+design and implementation section of this proposal mainly talks about the
+enforcement of the limit for maximum number of outstanding requests sent to
+an upstream cluster. See [Other configurations considered](#other-configurations-considered) 
+for rational.
+
+To match Envoy's implementation, circuit breaking limits will always be 
+enforced with certain defaults, even if the gRPC client receives no applicable 
+circuit breakers. Users can effectively turn circuit breaking off by setting
+thresholds to very high values, for example, to 
+`std::numeric_limits<uint32_t>::max()`.
+
+## Detailed design
+
+### Maintaining thresholds
+
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
+requests to hosts in the cluster that this policy is load balancing for. Make
+sure this threshold can be updated when receiving an update from the
+XdsClient cluster watch interface and the updated value can be dynamically 
+picked up by request limiting logic. By default, this value is set to 1024.
+
+### Limiting outstanding requests
+
+Each CDS LB policy uses a counter to count the number of requests sent through
+the _subchannels_ picked by its _Picker_, which is the output of the CDS LB
+policy. The counter has the same lifecycle as the CDS LB policy. In the _Pick_
+operation, the outputted Picker instance should includes the logic of checking 
+if the counter's value reaches the threshold before delegating the pick to the 
+Picker of downstream LB policies. If the threshold is reached, it returns
+a result that leads to the request being dropped with `UNAVAIABLE` status and
+never gets retried. Also this drop is recorded and aggregated to the 
+`total_dropped_requests` in the cluster-level stats to be reported in
```

Comment:
Currently, the drops are recorded in the EDS policy, not the CDS policy.  Does this mean we need to start reporting drops from both places?  Or can we coaleasce this somehow?

Reply:
My bad, this wasn't clearly mentioned in the design doc. But yes, we would need to be able to record drops in CDS policy. This drop stats is somewhat different from drops in EDS policy. Drops made in EDS policy are aggregated to explicit categories specified by DropOverload policies received in the EDS response. Drops due to hitting circuit breaking limits are not classified by categories, they are just accumulated to the [`total_dropped_requests`](https://github.com/envoyproxy/envoy/blob/10fed474642a90760b2a73f866087ef12d67e542/api/envoy/config/endpoint/v3/load_report.proto#L154-L155) of the `ClusterStats`.

Slight changes for managing recorded load stats might be necessary. Upon receiving an CDS update, the CDS policy tends to create a child policy based on the received eds_service_name. At this time, we should create a load recorder for this cluster:eds_service as well. Since load recorders are managed by the XdsClient, it shouldn't be a problem to do so. I can write up a section in the design doc to clearly describe the full flow of load stats recording and how its APIs look like if you want.

Reply:
If we do this in the CDS policy, then I think we should handle it by having the XdsClient provide a separate API for use by the CDS policy than the one it already provides for use by the EDS policy.  This is exactly analogous to the split we discussed as part of the xds policy refactoring to have the drop stats reported by the EDS policy and the actual request stats reported by the LRS policy.

Alternatively, we could handle this by moving the enforcement of circuit breaking down into the EDS policy instead, so that these drops can be reported at the same place as we report the other drops.  The only change we'd really need to make that work would be to add a field to the EDS policy's config indicating the circuit breaking `max_requests` limit.

Reply:
We should not move the enforcement of circuit breaking into the EDS policy, as an CDS policy may switch its child EDS policies when an updated `eds_service_name` is received. Requests sent through pickers outputted by all of this child EDS policies should be aggregated together and counted against the cluster-level threshold. 

We should just add a getter API on XdsClient for accessing the stats object. The stats object for each cluster:eds_service should have the same lifecycle as the corresponding EDS policy and have the CDS policy manages its lifecycle.

The workflow would be:
When the CDS policy receives an CDS update (includes information for eds_service_name, lrs_server_name), the stats object for this cluster:eds_service should be created along with instantiating the child EDS policy for this eds_service_name. The CDS policy manages to shutdown the old child EDS policy, it should destroy (un-reference) the stats object for it as well. Inside the EDS policy, it calls the getter API to access the stats object for recording drops if necessary. In this way, both CDS and EDS policies have access to the stats object. Moreover, having the CDS policy managing the lifecycle of the stats object seems to be more natural.

Reply:
I don't think it makes sense to try to tightly couple the lifetime of this object between the two policies, especially since they are not reporting the same data anyway.  I think it would be better to just have a separate object for the drop stats reported by the CDS policy, just like we do for the drop stats reported by the EDS policy vs. the request stats reported by the LRS policy.

Reply:
The drop stats recorded in both policy are eventually put together for load reporting. The `total_dropped_requests` in `ClusterStats` is the sum of drops recorded by the CDS policy and all of the drops recorded by EDS policies. We are just calliing different APIs of the stats object to record drops in both policies, but their data will still be put together. So it makes sense to use the same object recording both kinds of drops. We are bookkeeping a group of stats object with each contains the stats for a single cluster:eds_service. Its scope, per cluster:eds_service, is same as that for the EDS policy. The meaning is when the CDS policy creates its child EDS policies, it also creates objects to record stats for each of them.

Reply:
The fact that the stats are put together for load reporting is irrelevant.  That's also true of the stats reported by the EDS and LRS policies: they all go into the same LRS request.  But they get reported via separate objects.

I think it's better to have separate objects for reporting each type of data and then have the XdsClient merge the data when it constructs the load report.  That seems much cleaner to me than trying to tightly couple the lifetime of the object used by multiple LB policies.

Reply:
Alright, if you'd like a separate pair of add/remove APIs for stats object to be used by the CDS policy, that's fine. But I am more leaning towards leave this as an implementation detail and let implementors decide the approach based on how each language manages child policies in their LB policy implementations. As a design, we would just mention drops caused by circuit breakers should be recorded in `total_dropped_requests` for load reporting.